bin/node-template/node/src/command.rs

```diff
@@ -47,36 +47,36 @@ impl SubstrateCli for Cli {
 		2017
 	}
 
-	fn load_spec(&self, id: &str) -> Result<Box<dyn sc_service::ChainSpec>, String> {
-		Ok(match id {
-			"dev" => Box::new(chain_spec::development_config()?),
-			"" | "local" => Box::new(chain_spec::local_testnet_config()?),
-			path => Box::new(chain_spec::ChainSpec::from_json_file(
-				std::path::PathBuf::from(path),
-			)?),
-		})
-	}
-
 	fn native_runtime_version(_: &Box<dyn ChainSpec>) -> &'static RuntimeVersion {
 		&node_template_runtime::VERSION
 	}
 }
 
+fn spec_factory(id: String) -> Result<Box<dyn ChainSpec>, String> {
+	Ok(match &*id {
+		"dev" => Box::new(chain_spec::development_config()?),
+		"" | "local" => Box::new(chain_spec::local_testnet_config()?),
+		path => Box::new(chain_spec::ChainSpec::from_json_file(
+			std::path::PathBuf::from(path),
+		)?),
+	})
+}
```

Comment:
Why are you moving that out again? It wasn't nice inside SubstrateCli?

Reply:
> It wasn't nice inside SubstrateCli?

unfortunately, having the spec_factory in the cli complicates the API for paritytech/substrate-test-runner, where users would also have to provide `impl SubstrateCli` simply because we want the spec_factory. Splitting it out keeps our API lean and simple.

Reply:
Ah it's for this? https://github.com/paritytech/substrate-test-runner/pull/2

Reply:
ok ok

Reply:
Is it related in any way with this? https://github.com/paritytech/substrate/pull/6651

Are you trying to run nodes inside tests?

Reply:
@seunlanlege I don't see the full picture yet so bare with me I might be completely wrong. If you can provide me documentation and more information on the project it would help me understand better the big picture.

Wild guess: I don't think you need sc-cli at all. sc-cli is for building clis, here you are building a library, you should be using sc-service and use the sc_service::Configuration object directly. It's sc-service that is the glue code and the center of everything, not sc-cli. You can't just pick the elements you want from sc-cli without applying to the full logic of sc-cli's usage.

Example: if you pass the arguments of the cli by code, I'm confident that this is wrong. sc-cli is really only meant for building a binary and the arguments are passed by a user.

```rust
impl<Runtime> InternalNode<Runtime> {
	pub fn builder() -> InternalNodeBuilder<Runtime> {
		InternalNodeBuilder::new()
	}

	pub fn new(logs: Logger, cli: &[String]) -> Self {
		let cli = node_cli::Cli::from_iter(cli.iter());
...
```

Reply:
(In other words: sc-cli must be a dependency of a bin crate, not a lib crate)

Reply:
@cecton Please show them your code you have done for the polkadot test service. Which solves this problem AFAIK.

Reply:
https://github.com/paritytech/polkadot/blob/master/node/test-service/src/lib.rs

@seunlanlege This is a testing crate for polkadot. It uses sc-service directly. I just pass the chain-spec I want, tweak the sc_service::Configuration object the way I want, and that's how I can run a node. The nodes can communicate with each other using memory network so it doesn't open a port on the machine and I don't need to know the port number.

The PR #6651 is meant to improve testing nodes by providing an async runtime for the test and a timeout.

The PR #6555 is also meant to improve testing nodes by providing a "wait_for_blocks" function (to check that blocks are being created) and "send_transaction" to arbitrarily send a transaction to the node.